presto-hudi/src/main/java/com/facebook/presto/hudi/HudiSchemaEvolutionUtils.java

```diff
@@ -0,0 +1,192 @@
+import org.apache.hudi.common.table.timeline.HoodieInstant;
+import org.apache.hudi.common.table.timeline.HoodieTimeline;
+import org.apache.hudi.common.util.InternalSchemaCache;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.internal.schema.InternalSchema;
+import org.apache.hudi.internal.schema.Type;
+import org.apache.hudi.internal.schema.Types;
+import org.apache.hudi.internal.schema.action.InternalSchemaMerger;
+import org.apache.hudi.internal.schema.utils.InternalSchemaUtils;
+import org.apache.hudi.internal.schema.utils.SerDeHelper;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.hive.HiveType.toHiveType;
+
+public final class HudiSchemaEvolutionUtils
+{
+    private static final Logger log = Logger.get(HudiSchemaEvolutionUtils.class);
+
+    private HudiSchemaEvolutionUtils()
+    {
+    }
+
+    public static SchemaEvolutionContext createSchemaEvolutionContext(HoodieTableMetaClient metaClient)
+    {
+        // no need to do schema evolution for mor table, since hudi kernel will do it.
+        if (metaClient.getTableType() == HoodieTableType.MERGE_ON_READ || metaClient == null) {
+            return new SchemaEvolutionContext("", "");
+        }
+
+        try {
+            TableSchemaResolver schemaUtil = new TableSchemaResolver(metaClient);
+            String internalSchema = schemaUtil.getTableInternalSchemaFromCommitMetadata().map(SerDeHelper::toJson).orElse("");
+            HoodieTimeline hoodieTimeline = metaClient.getCommitsAndCompactionTimeline().filterCompletedInstants();
+            String validCommits = hoodieTimeline.getInstants().map(HoodieInstant::getFileName).collect(Collectors.joining(","));
+            return new SchemaEvolutionContext(internalSchema, validCommits);
+        }
+        catch (Exception e) {
+            log.warn(String.format("failed to get internal Schema from hudi table：%s , fallback to original logical", metaClient.getBasePathV2()), e);
+        }
+        return new SchemaEvolutionContext("", "");
+    }
+
+    public static Pair<List<HudiColumnHandle>, Map<String, HiveType>> doEvolution(HudiSplit hudiSplit, List<HudiColumnHandle> oldColumnHandle, String tablePath, Configuration hadoopConf)
+    {
+        SchemaEvolutionContext schemaEvolutionContext = hudiSplit.getSchemaEvolutionContext();
+        InternalSchema internalSchema = SerDeHelper.fromJson(schemaEvolutionContext.getRequiredSchema()).orElse(InternalSchema.getEmptyInternalSchema());
+        if (internalSchema.isEmptySchema() || !hudiSplit.getBaseFile().isPresent()) {
+            return Pair.of(oldColumnHandle, ImmutableMap.of());
+        }
+        // prune internalSchema: columns prune
+        InternalSchema prunedSchema = InternalSchemaUtils.pruneInternalSchema(internalSchema, oldColumnHandle.stream().map(HudiColumnHandle::getName).collect(Collectors.toList()));
```

Comment:
good question.
At present, hudi cannot guarantee that the metadata in hive is consistent with the metadata of the current table, and users need to ensure that. this is a big problem.

In this case, we it will be better to throw an exception directly and prompt the user that the metadata information of the current hive table is inconsistent with the data information of the hudi table 

WDYT？

Reply:
We should throw an error as metastore is behind hudi table and needs to be synced again.